client/fetch_tagged_results_accumulator.go

```diff
@@ -0,0 +1,366 @@
+	AsEncodingSeriesIterators(limit int, pools fetchTaggedPools) (
+		iters encoding.SeriesIterators, exhaustive bool, err error)
+
+	// AsIndexQueryResults converts the underlying response into an index.QueryResults
+	AsIndexQueryResults(limit int, pools fetchTaggedPools) (index.QueryResults, error)
+
+	// Clear clears any internal state, and releases references the object
+	// has to external entities. It does not release the object itself.
+	Clear()
+}
+
+type fetchTaggedResultAccumulatorOpts struct {
+	host     topology.Host
+	response *rpc.FetchTaggedResult_
+}
+
+// make the compiler ensure the concrete type `&fetchTaggedResultAccumulator{}` implements
+// the `fetchResultsAccumulator` interface.
+var _ fetchResultsAccumulator = &fetchTaggedResultAccumulator{}
+
+type fetchTaggedResultAccumulator struct {
+	// NB(prateek): a fetchTagged request requires we fan out to each shard in the
+	// topology. As a result, we model the response consistency per shard.
+	// Length of this slice == 1 + max shard id in topology
+	shardResults     []fetchTaggedShardConsistencyResult
+	numHostsPending  int32
+	numShardsPending int32
+
+	errors     xerrors.Errors
+	responses  fetchTaggedIDResults
+	exhaustive bool
+
+	startTime        time.Time
+	endTime          time.Time
+	majority         int
+	consistencyLevel topology.ReadConsistencyLevel
+	topoMap          topology.Map
+}
+
+type fetchTaggedShardConsistencyResult struct {
+	enqueued int8
+	success  int8
+	errors   int8
+	done     bool
+}
+
+func (accum *fetchTaggedResultAccumulator) Add(
+	opts fetchTaggedResultAccumulatorOpts,
+	resultErr error,
+) (bool, error) {
+	host := opts.host
+	response := opts.response
+
+	accum.numHostsPending--
+	if resultErr != nil {
+		accum.errors = append(accum.errors, xerrors.NewRenamedError(resultErr,
+			fmt.Errorf("error fetching tagged from host %s", host.ID())))
+	} else {
+		accum.exhaustive = accum.exhaustive && response.Exhaustive
+		for _, elem := range response.Elements {
+			accum.responses = append(accum.responses, elem)
+		}
+	}
+
+	hostShardSet, ok := accum.topoMap.LookupHostShardSet(host.ID())
+	if !ok {
+		// should never happen, as we've taken a reference to the
+		// topology when beginning the request, and the var is immutable.
+		doneAccumulating := true
+		err := fmt.Errorf(
+			"[invariant violated] missing host shard in fetchState completionFn: %s", host.ID())
+		return doneAccumulating, xerrors.NewNonRetryableError(err)
+	}
+
+	for _, hs := range hostShardSet.ShardSet().All() {
+		if hs.State() != shard.Available {
+			// Currently, we only accept responses from shard's which are available
+			// NB: as a possible enhancement, we could accept a response from
+			// a shard that's not available if we tracked response pairs from
+			// a LEAVING+INITIALIZING shard; this would help during node replaces.
+			continue
+		}
+
+		shardID := int(hs.ID())
+		shardResult := accum.shardResults[shardID]
```

Comment:
Is it worth bounds checking here?

Reply:
Don't need to, we explicitly ensure `shardResults` is big enough in the `Reset()` method.

Reply:
Yeah fair, topoMap is captured already.